analyzers/src/SonarAnalyzer.Common/Rules/HardcodedIpAddressBase.cs

```diff
@@ -20,74 +20,130 @@
 
 using System;
 using System.Collections.Generic;
+using System.Collections.Immutable;
 using System.Linq;
 using System.Net;
 using System.Net.Sockets;
+using System.Text.RegularExpressions;
 using Microsoft.CodeAnalysis;
 using Microsoft.CodeAnalysis.Diagnostics;
 using SonarAnalyzer.Common;
 using SonarAnalyzer.Helpers;
 
 namespace SonarAnalyzer.Rules
 {
-    public abstract class HardcodedIpAddressBase<TLiteralExpression> : HotspotDiagnosticAnalyzer
+    public abstract class HardcodedIpAddressBase<TSyntaxKind, TLiteralExpression> : HotspotDiagnosticAnalyzer
+        where TSyntaxKind : struct
         where TLiteralExpression : SyntaxNode
     {
+        private static bool IsBroadcast(string ip) =>
+            ip.Equals("255.255.255.255", StringComparison.InvariantCultureIgnoreCase);
+
+        private static bool DoesItLookLikeObjectIdentifier(string s) =>
+            s.StartsWith("2.5.");
+
         protected const string DiagnosticId = "S1313";
+
         protected const string MessageFormat = "Make sure using this hardcoded IP address '{0}' is safe here.";
 
-        private static readonly ISet<string> IgnoredVariableNames =
+        private const int NumOfIPv4AddressParts = 4;
+
+        private readonly ISet<string> ignoredVariableNames =
             new HashSet<string>
             {
                 "VERSION",
                 "ASSEMBLY",
             };
 
+        private readonly Regex ipv6LoopbackPattern = new Regex("^(?:0*:)*?:?0*1$", RegexOptions.Compiled);
```

Comment:
`^(?:0*:)*?:?0*1$` Original
`^(0*:)*?:?0*1$` Why using noncapturing groups when we don't capture anything? 
`^(0*:)*:?0*1$` As we fix `begin`, `end` and have defined content, this should behave the same

But it still matches many inputs that are not IPv6 address. Like `:::::1`, `::000001`, `0::0::1` and so on. Let's get rid of this if possible with the other comment about `IPAddress.IsLoopback`